src/Analyzers/CSharp/Analyzers/RemoveUnnecessarySuppressions/UnnecessaryNullableWarningSuppressionsUtilities.cs

```diff
@@ -0,0 +1,120 @@
+            return false;
+
+        var semanticDiagnostics = semanticModel.GetDiagnostics(postfixUnary.Span, cancellationToken);
+        if (ContainsErrorOrWarning(semanticDiagnostics))
+            return false;
+
+        return true;
+    }
+
+    public static void AddUnnecessaryNodes(
+        SemanticModel semanticModel,
+        ArrayBuilder<PostfixUnaryExpressionSyntax> result,
+        CancellationToken cancellationToken)
+    {
+        using var _1 = ArrayBuilder<PostfixUnaryExpressionSyntax>.GetInstance(out var nodesToCheck);
+
+        var originalTree = semanticModel.SyntaxTree;
+        var originalRoot = originalTree.GetRoot(cancellationToken);
+
+        foreach (var existingNode in originalRoot.DescendantNodes())
+        {
+            if (existingNode is PostfixUnaryExpressionSyntax postfixUnary &&
+                ShouldAnalyzeNode(semanticModel, postfixUnary, cancellationToken))
+            {
+                nodesToCheck.Add(postfixUnary);
+            }
+        }
+
+        if (nodesToCheck.IsEmpty)
+            return;
+
+        using var _2 = PooledDictionary<PostfixUnaryExpressionSyntax, SyntaxAnnotation>.GetInstance(out var nodeToAnnotation);
+        foreach (var node in nodesToCheck)
+            nodeToAnnotation[node] = new SyntaxAnnotation();
+
+        ForkSemanticModelAndCheckNodes(semanticModel, nodesToCheck, nodeToAnnotation, result, cancellationToken);
+    }
+
+    private static void ForkSemanticModelAndCheckNodes(
+        SemanticModel semanticModel,
+        ArrayBuilder<PostfixUnaryExpressionSyntax> nodesToCheck,
+        PooledDictionary<PostfixUnaryExpressionSyntax, SyntaxAnnotation> nodeToAnnotation,
+        ArrayBuilder<PostfixUnaryExpressionSyntax> result,
+        CancellationToken cancellationToken)
+    {
+        var originalTree = semanticModel.SyntaxTree;
+        var originalRoot = originalTree.GetRoot(cancellationToken);
+
+        var updatedTree = originalTree.WithRootAndOptions(
+            originalRoot.ReplaceNodes(
+                nodesToCheck,
+                (original, current) => current.Operand.WithAdditionalAnnotations(nodeToAnnotation[original])),
+            originalTree.Options);
+        var updateRoot = updatedTree.GetRoot(cancellationToken);
+        var updatedCompilation = semanticModel.Compilation
+            .ReplaceSyntaxTree(originalTree, updatedTree)
+            .WithOptions(semanticModel.Compilation.Options.WithSpecificDiagnosticOptions([]));
+        var updatedSemanticModel = updatedCompilation.GetSemanticModel(updatedTree);
+
+        foreach (var (node, annotation) in nodeToAnnotation)
+        {
+            var updatedNode = updateRoot.GetAnnotatedNodes(annotation).Single();
+            var span = GetSpanToCheck(updatedNode);
+            if (span is null)
+                continue;
+
+            var updatedDiagnostics = updatedSemanticModel.GetDiagnostics(span, cancellationToken);
+            if (ContainsErrorOrWarning(updatedDiagnostics))
```

Comment:
It looks like the rough solution here is:

1. track all the `expr!` nodes from the original tree.
2. make a derived tree which changes all `expr!` to expr`
3. get a semantic model for the derived tree, and diagnostics for the method/subtree/whatever we are analyzing
4. for each `expr` which was previously suppressed, see if there is a diagnostic on it
5. mark `expr`s which don't have a diagnostic as having an unnecessary suppression in the user's code.

Is that right? Seems good, I think. I am curious if this is similar to how the "unnecessary `#pragma`" analysis works.